mitiq/interface/mitiq_qibo/conversions.py

```diff
@@ -0,0 +1,531 @@
+    """Decomposes RYY gate to Cirq known gates.
+
+    Args:
+        gate: RYY gate to decompose.
+
+    Returns:
+        List with gates that has the same effect as applying the original gate.
+    """
+    q0, q1 = gate.init_args
+    theta = gate.parameters[0]
+    decomp_gate = [
+        gates.RX(q0, pi / 2),
+        gates.RX(q1, pi / 2),
+        gates.CNOT(q0, q1),
+        gates.RZ(q1, theta),
+        gates.CNOT(q0, q1),
+        gates.RX(q0, -pi / 2),
+        gates.RX(q1, -pi / 2),
+    ]
+    return decomp_gate
+
+
+def rzz_decomp(gate: gates.RZZ) -> List[Gate]:
+    """Decomposes RZZ gate to Cirq known gates.
+
+    Args:
+        gate: RZZ gate to decompose.
+
+    Returns:
+        List with gates that has the same effect as applying the original gate.
+    """
+    q0, q1 = gate.init_args
+    theta = gate.parameters[0]
+    decomp_gate = [
+        gates.CNOT(q0, q1),
+        gates.RZ(q1, theta),
+        gates.CNOT(q0, q1),
+    ]
+    return decomp_gate
+
+
+GATES_TO_DECOMPOSE = {
+    "crx": crx_decomp,
+    "cry": cry_decomp,
+    "crz": crz_decomp,
+    "cu1": cu1_decomp,
+    "cu3": cu3_decomp,
+    "iswap": iswap_decomp,
+    "fswap": fswap_decomp,
+    "rxx": rxx_decomp,
+    "ryy": ryy_decomp,
+    "rzz": rzz_decomp,
+}
+
+
+class UnsupportedCirqCircuitError(Exception):
+    pass
+
+
+class UnsupportedQiboCircuitError(Exception):
+    pass
+
+
+def decompose_qibo_circuit(qibo_circuit: QiboCircuit) -> QiboCircuit:
+    """Returns a QiboCircuit circuit equivalent to the input QiboCircuit
+    with all unknown by cirq gates decomposed.
+
+    Args:
+        qibo_circuit: QiboCircuit to decompose.
+
+    Returns:
+        Decomposed QiboCircuit
+    """
+    decomp_circuit = QiboCircuit(qibo_circuit.nqubits)
+    for gate in qibo_circuit.queue:
+        if gate.name in GATES_TO_DECOMPOSE:
+            function = GATES_TO_DECOMPOSE[gate.name]
+            decomposed_gate = function(gate)
+            decomp_circuit.add(decomposed_gate)
+        else:
+            decomp_circuit.add(gate)
+
+    return decomp_circuit
+
+
+def from_qibo(qibo_circuit: QiboCircuit) -> Circuit:
+    """Returns a Cirq circuit equivalent to the input QiboCircuit.
+
+    Args:
+        qibo_circuit: QiboCircuit to convert to a Cirq circuit.
+
+    Returns:
+        Cirq circuit representation equivalent to the input QiboCircuit.
+    """
+    for measurement in qibo_circuit.measurements:
+        reg_name = measurement.register_name
+        reg_qubits = measurement.target_qubits
+        if not reg_name.islower():
+            raise UnsupportedQiboCircuitError(
+                f"OpenQASM does not support capital letters in register names but {reg_name} was used."
+            )
+    for gate in qibo_circuit.queue:
+        if isinstance(gate, gates.M):
+            continue
+        if gate.is_controlled_by:
+            raise UnsupportedQiboCircuitError(
+                "OpenQASM does not support multi-controlled gates."
+            )
```

Comment:
From checking out the code of `controlled_by` (which `is_controlled_by` refers to), it seems like you can create a singly-controlled gate using `controlled_by`. Does OpenQASM support arbitrary singly-controlled gates?

Reply:
It appears so: https://openqasm.com/language/gates.html .

Reply:
@francescsabater can you comment on why this error here is needed? From the looks of the link Misty provided it looks like multi-controlled gates should be possible. Whether they are actually implemented is another story, but it would be good to know what the situation is.